superset-frontend/src/dashboard/components/FiltersBadge/selectors.ts

```diff
@@ -0,0 +1,146 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import { isNil, get } from 'lodash';
+import { getChartIdsInFilterScope } from '../../util/activeDashboardFilters';
+import { TIME_FILTER_MAP } from '../../../visualizations/FilterBox/FilterBox';
+
+export const UNSET = 'UNSET';
+export const APPLIED = 'APPLIED';
+export const INCOMPATIBLE = 'INCOMPATIBLE';
+
+const TIME_GRANULARITY_FIELDS = new Set([
+  TIME_FILTER_MAP.granularity,
+  TIME_FILTER_MAP.time_grain_sqla,
+]);
+
+type Filter = {
+  chartId: number;
+  columns: { [key: string]: string | string[] };
+  scopes: { [key: string]: any };
+  labels: { [key: string]: string };
+  isDateFilter: boolean;
+  directPathToFilter: string[];
+  datasourceId: string;
+};
+
+type Datasource = {
+  time_grain_sqla?: [string, string][];
+  granularity?: [string, string][];
+};
```

Comment:
can probably replace with `import { DatasourceMeta } from '@superset-ui/chart-controls'`

Reply:
Nice, thanks

Reply:
that type doesn't include `granularity`. It probably should, right?

Reply:
Also, the type for `time_grain_sqla` appears to be incomplete. It's `string` and should be `[string, string][]` (or maybe the union of those?)

Reply:
Hmm, `@superset-ui/chart-controls@v15.0.3` (which is in the `master` branch) should have made `time_grain_sqla` `[string, string][]`: https://github.com/apache-superset/superset-ui/blob/0f11940d8201feea3c5f5adee82aef68a9d201f3/packages/superset-ui-chart-controls/src/types.ts#L72

Based on
https://github.com/apache/incubator-superset/blob/fb6f14c4693c5764cc2a151c27746386b190eeae/superset/connectors/sqla/models.py#L661-L664

I believe it is only `[string, string][]` for `Datasource`.

As for `granularity`, feel free to add it in `superset-ui/chart-controls` if it's not too much hassle for you.

There is also `import { Datasource } from '@superset/core'`, which is missing both properties. We should probably consolidate these two instead of adding yet another new type.